tools/syz-trace2syz/proggen/proggen.go

```diff
@@ -0,0 +1,613 @@
+	case *parser.PointerType:
+		// This can be triggered by the following:
+		// 2435  connect(3, {sa_family=0x2f ,..., 16)
+		return prog.MakeConstArg(syzType, a.Address)
+	default:
+		log.Fatalf("Cannot convert Strace Type: %s to Const Type", traceType.Name())
+	}
+	return nil
+}
+
+func genResource(syzType *prog.ResourceType, traceType parser.IrType, ctx *Context) prog.Arg {
+	if syzType.Dir() == prog.DirOut {
+		log.Logf(2, "Resource returned by call argument: %s", traceType.String())
+		res := prog.MakeResultArg(syzType, nil, syzType.Default())
+		ctx.ReturnCache.cache(syzType, traceType, res)
+		return res
+	}
+	switch a := traceType.(type) {
+	case parser.Expression:
+		val := a.Eval(ctx.Target)
+		if arg := ctx.ReturnCache.get(syzType, traceType); arg != nil {
+			res := prog.MakeResultArg(syzType, arg.(*prog.ResultArg), syzType.Default())
+			return res
+		}
+		// May get a resource type like ifindex which is returned by ioctl$SIOCGIFINDEX
+		// but strace converts the value to a call type like if_nametoindex which is hard to evaluate
+		if arg := ctx.ReturnCache.findMatchingType(syzType, traceType); arg != nil {
+			res := prog.MakeResultArg(syzType, arg.(*prog.ResultArg), syzType.Default())
+			return res
+		}
+		res := prog.MakeResultArg(syzType, nil, val)
+		return res
+	case *parser.Field:
+		return genResource(syzType, a.Val, ctx)
+	case *parser.PointerType:
+		return prog.MakeResultArg(syzType, nil, 0)
+	case *parser.GroupType:
+		if len(a.Elems) == 1 {
+			// For example: 5028  ioctl(3, SIOCSPGRP, [0])          = 0
+			// last argument is a pointer to a resource. Strace will output a pointer to
+			// a number x as [x]. When we parse we cannot distinguish this case from a pointer to a pointer.
+			// It only matters if the value is resource AFAWCT
+			res := prog.MakeResultArg(syzType, nil, syzType.Default())
+			ctx.ReturnCache.cache(syzType, a.Elems[0], res)
+			return res
+		}
+		log.Fatalf("Generating resource type from GroupType with %d elements", len(a.Elems))
+	default:
+		log.Fatalf("Resource Type only supports Expression")
+	}
+	return nil
+}
+
+func parseProc(syzType *prog.ProcType, traceType parser.IrType, ctx *Context) prog.Arg {
+	if syzType.Dir() == prog.DirOut {
+		return genDefaultArg(syzType, ctx)
+	}
+	switch a := traceType.(type) {
+	case parser.Expression:
+		val := a.Eval(ctx.Target)
+		if val >= syzType.ValuesPerProc {
+			return prog.MakeConstArg(syzType, syzType.ValuesPerProc-1)
+		}
+		return prog.MakeConstArg(syzType, val)
+	case *parser.Field:
+		return genArgs(syzType, a.Val, ctx)
+	case *parser.BufferType:
+		// Again probably an error case
+		// Something like the following will trigger this
+		// bind(3, {sa_family=AF_INET, sa_data="\xac"}, 3) = -1 EINVAL(Invalid argument)
+		return genDefaultArg(syzType, ctx)
+	default:
+		log.Fatalf("Unsupported Type for Proc: %#v\n", traceType)
+	}
+	return nil
+}
+
+func genDefaultArg(syzType prog.Type, ctx *Context) prog.Arg {
```

Comment:
Can't we just use prog.DefaultArg instead of this.
For StructType and UnionType this simply copy-pastes code from prog, for most remaining types it calls prog.DefaultArg.
I see a difference for PtrType, but if we are here then we already gave up on generating good value, so we could as well just use DefaultArg for PtrType too.
I don't fully understand what happens for ResourceType. But I suspect it may be the root cause of the bug that I described before where fd=-1 transforms into result of socket call. No, if we use default resource value, we don't need to cache it and reuse in future (which one if there are several calls?). If fd=-1 is used later, then we just use fd=-1 rather than result of some preceding call.
So if I am not terribly mistaken, we can just replace this whole function with prog.DefaultArg.

Reply:
The problem is that for calls like pipe() or socketpair() their descriptors get lost. 

The reason for the bug is here: https://github.com/shankarapailoor/syzkaller/blob/master/tools/syz-trace2syz/proggen/proggen.go#L474, which I added a few days ago while trying to parse traces of syzkaller generated programs.

Reply:
So we don't cache -1 results. We only cache positive results.

Reply:
Ok so I agree that this function can be removed. I don't think prog.DefaultArg can be called only in genArgs for now.